### DISCUSSION TITLE
Explore faster serialization options

### DISCUSSION THREAD

**Discussion by philippjfr** [bokeh/bokeh#11376] [Internals and design (Q&A)]
Recently Plotly switched to [orjson](https://github.com/ijl/orjson) for their JSON serialization implementation which they claim brought significant speedups. We could explore whether it would be easy to swap the JSON serializer Bokeh uses after performing some profiling to determine whether this step is an actual bottleneck.

Some tasks related to this:

1. Profile which step during serialization is actually slow
2. Explore how hard it is to make the json serializer swappable

Related topics:

- Explore a non-websocket based server (this is where faster JSON serialization could be very helpful)

**Comment by mattpap** [comment]
This library is written in rust, so I guess the bulk of the speedup comes from using a compiled language and having an implementation that's focused only on performance. This raises a question about extensibility, given the richness of types our serialization scheme supports. I think it may still be useful to consider this as an alternative to stdlib, but only for the final step of transformation to (a) JSON (string). Currently we need to improve the earlier stages, e.g.:

1. Don't store data in the primary JSON structure, even in standalone embedding (e.g. binary base64 encode in separate `<script>` tags).
2. Don't do `loads(dumps(obj))` to get the intermediate JSON-like structure (that's done due to limitations of extensibility of the standard JSON library).

**Reply by bryevdv** [reply]
> Don't store data in the primary JSON structure, even in standalone embedding

I am very sympathetic to this goal, but I do think we need to keep UX at the forefront of any decisions. Users already often struggle with our embedding options, so if we are going to make things more complicated (from their perspective) we need to make sure to do two things:

* find ways to offset the complexity by simplifying elsewhere (UX point of view, not ours)
* very clearly spell out *user* benefits, e.g. if we initialize from data in separate script tag, could we delete the script tag from the DOM after init? Some users might appreciate that. Alternatively could fixed model-content now be easily "parametrizable" or "reusable" over different data script tags? Could the data be loaded entirely separately e.g. via an endpoint?
